xds/internal/balancer/edsbalancer/balancergroup_test.go

```diff
@@ -450,3 +457,104 @@ func TestBalancerGroup_LoadReport(t *testing.T) {
+// - b1, weight 2, backends [0,1]
+// - b2, weight 1, backends [2,3]
+// Start the balancer group and check behavior.
+//
+// Close the balancer group, call add/remove/change weight/change address.
+// - b2, weight 3, backends [0,3]
+// - b3, weight 1, backends [1,2]
+// Start the balancer group again and check for behavior.
+func TestBalancerGroup_start_close(t *testing.T) {
+	cc := newTestClientConn(t)
+	bg := newBalancerGroup(cc, nil)
+
+	// Add two balancers to group and send two resolved addresses to both
+	// balancers.
+	bg.add(testBalancerIDs[0], 2, rrBuilder)
+	bg.handleResolvedAddrs(testBalancerIDs[0], testBackendAddrs[0:2])
+	bg.add(testBalancerIDs[1], 1, rrBuilder)
+	bg.handleResolvedAddrs(testBalancerIDs[1], testBackendAddrs[2:4])
+
+	bg.start()
+
+	m1 := make(map[resolver.Address]balancer.SubConn)
+	for i := 0; i < 4; i++ {
+		addrs := <-cc.newSubConnAddrsCh
+		sc := <-cc.newSubConnCh
+		m1[addrs[0]] = sc
+		bg.handleSubConnStateChange(sc, connectivity.Connecting)
+		bg.handleSubConnStateChange(sc, connectivity.Ready)
+	}
+
+	// Test roundrobin on the last picker.
+	p1 := <-cc.newPickerCh
+	want := []balancer.SubConn{
+		m1[testBackendAddrs[0]], m1[testBackendAddrs[0]],
+		m1[testBackendAddrs[1]], m1[testBackendAddrs[1]],
+		m1[testBackendAddrs[2]], m1[testBackendAddrs[3]],
+	}
+	if err := isRoundRobin(want, func() balancer.SubConn {
+		sc, _, _ := p1.Pick(context.Background(), balancer.PickOptions{})
+		return sc
+	}); err != nil {
+		t.Fatalf("want %v, got %v", want, err)
+	}
+
+	bg.close()
+	for i := 0; i < 4; i++ {
+		bg.handleSubConnStateChange(<-cc.removeSubConnCh, connectivity.Shutdown)
+	}
+
+	// Add b3, weight 1, backends [1,2].
+	bg.add(testBalancerIDs[2], 1, rrBuilder)
+	bg.handleResolvedAddrs(testBalancerIDs[2], testBackendAddrs[1:3])
+
+	// Remove b1.
+	bg.remove(testBalancerIDs[0])
+
+	// Update b2 to weight 3, backends [0,3].
+	bg.changeWeight(testBalancerIDs[1], 3)
+	bg.handleResolvedAddrs(testBalancerIDs[1], append([]resolver.Address(nil), testBackendAddrs[0], testBackendAddrs[3]))
+
+	bg.start()
+
+	m2 := make(map[resolver.Address]balancer.SubConn)
+	for i := 0; i < 4; i++ {
+		addrs := <-cc.newSubConnAddrsCh
+		sc := <-cc.newSubConnCh
+		m2[addrs[0]] = sc
+		bg.handleSubConnStateChange(sc, connectivity.Connecting)
+		bg.handleSubConnStateChange(sc, connectivity.Ready)
+	}
+
+	// Test roundrobin on the last picker.
+	p2 := <-cc.newPickerCh
+	want = []balancer.SubConn{
+		m2[testBackendAddrs[0]], m2[testBackendAddrs[0]], m2[testBackendAddrs[0]],
+		m2[testBackendAddrs[3]], m2[testBackendAddrs[3]], m2[testBackendAddrs[3]],
+		m2[testBackendAddrs[1]], m2[testBackendAddrs[2]],
+	}
+	if err := isRoundRobin(want, func() balancer.SubConn {
+		sc, _, _ := p2.Pick(context.Background(), balancer.PickOptions{})
+		return sc
+	}); err != nil {
+		t.Fatalf("want %v, got %v", want, err)
+	}
+}
+
+func TestBalancerGroup_start_close_deadlock(t *testing.T) {
```

Comment:
I don't seem to understand what this test is testing.

Reply:
The test balancer updates picker in line when it receives an address update.
Comment added.